Framework/Algorithms/src/Q1DWeighted.cpp

```diff
@@ -154,199 +139,228 @@ void Q1DWeighted::exec() {
-    std::vector<std::vector<double>> wedgeLambdaIqErr(
-        nWedges, std::vector<double>(sizeOut - 1, 0.0));
-    std::vector<std::vector<double>> wedgeXNorm(
-        nWedges, std::vector<double>(sizeOut - 1, 0.0));
-
-    for (int i = 0; i < numSpec; i++) {
-      if (!spectrumInfo.hasDetectors(i)) {
-        g_log.warning() << "Workspace index " << i
-                        << " has no detector assigned to it - discarding\n";
+    // loop over lambda bins
+    for (size_t j = 0; j < nLambda; ++j) {
+
+      // skip if the bin is masked
+      if (std::binary_search(maskedBins.cbegin(), maskedBins.cend(), j)) {
         continue;
       }
-      // Skip if we have a monitor or if the detector is masked.
-      if (spectrumInfo.isMonitor(i) || spectrumInfo.isMasked(i))
-        continue;
 
-      // Get the current spectrum for both input workspaces
-      auto &XIn = inputWS->x(i);
-      auto &YIn = inputWS->y(i);
-      auto &EIn = inputWS->e(i);
+      const double wavelength = (XIn[j] + XIn[j + 1]) / 2.;
 
-      // Each pixel is sub-divided in the number of pixels given as input
-      // parameter (NPixelDivision)
-      for (int isub = 0; isub < nSubPixels * nSubPixels; isub++) {
-        // Find the position offset for this sub-pixel in real space
-        double subY = pixelSizeY *
-                      ((isub % nSubPixels) - (nSubPixels - 1.0) / 2.0) /
-                      nSubPixels;
-        double subX = pixelSizeX *
-                      (floor(static_cast<double>(isub) / nSubPixels) -
-                       (nSubPixels - 1.0) * 0.5) /
-                      nSubPixels;
-
-        // Find the position of this sub-pixel in real space and compute Q
-        // For reference - in the case where we don't use sub-pixels, simply
-        // use:
-        //     double sinTheta = sin( spectrumInfo.twoTheta(i)/2.0 );
-        V3D pos = spectrumInfo.position(i) - V3D(subX, subY, 0.0) - samplePos;
-        double sinTheta = sin(0.5 * pos.angle(beamLine));
-        double factor = fmp * sinTheta;
-        double q = factor * 2.0 / (XIn[j] + XIn[j + 1]);
-        int iq = 0;
-
-        // Bin assignment depends on whether we have log or linear bins
-        if (binParams.size() == 3) {
-          if (binParams[1] > 0.0) {
-            iq = static_cast<int>(floor((q - binParams[0]) / binParams[1]));
-          } else {
-            iq = static_cast<int>(
-                floor(log(q / binParams[0]) / log(1.0 - binParams[1])));
-          }
-          // If we got a more complicated binning, find the q bin the slow way
-        } else {
-          for (int iQbin = 0; iQbin < static_cast<int>(XOut.size()) - 1;
-               iQbin++) {
-            if (q >= XOut[iQbin] && q < XOut[(iQbin + 1)]) {
-              iq = iQbin;
-              break;
-            }
-          }
-        }
+      V3D correction;
+      if (correctGravity) {
+        correction.setY(gravityHelper.gravitationalDrop(wavelength));
```

Comment:
Assumes that `Y` is the up direction? Should this use some information from the instrument instead?